desc/experimental/_terpsichore.py

```diff
@@ -0,0 +1,872 @@
+    """Run TERPSICHORE."""
+    stdout_path = os.path.join(dir, "stdout.terps")
+    stderr_path = os.path.join(dir, "stderr.terps")
+
+    fout = open(stdout_path, "w")
+    ferr = open(stderr_path, "w")
+
+    cmd = exec + " < " + input + " " + wout
+    terps_subprocess = subprocess.run(
+        cmd, cwd=dir, shell=True, stdout=fout, stderr=ferr
+    )
+
+    run_time = 0.0
+    tic = time.perf_counter()
+    while not _is_terps_complete(
+        path=stdout_path,
+        run_time=run_time,
+        stop_time=stop_time,
+        terps_subprocess=terps_subprocess,
+    ):
+        time.sleep(sleep_time)
+        toc = time.perf_counter()
+        run_time = toc - tic
+
+    fout.close()
+    ferr.close()
+
+
+def _is_terps_complete(path, run_time, stop_time, terps_subprocess):
+    """Check if TERPSICHORE is finished running."""
+    if not os.path.exists(path):
+        return False
+
+    errorif(
+        run_time > stop_time, RuntimeError, "TERPS was unable to find a growth rate!"
+    )
+
+    file = open(path)
+    terps_output = file.read()
+    if "GROWTH RATE" in terps_output:
+        if terps_subprocess.returncode is None:
+            terps_subprocess.terminate()
+        file.close()
+        return True
+    else:
+        file.close()
+        return False
+
+
+def _read_terps_output(path, scalar, surfs):
+    """Read TERPSICHORE output file and return the growth rate / dW values."""
+    errorif(
+        not os.path.exists(path),
+        RuntimeError,
+        "TERPSICHORE fort.16 output file not found!",
+    )
+
+    file = open(path)
+    line_num = 0
+    growth_rate = []
+    delta_W = []
+    for k, line in enumerate(file):
+        index = line.find("GROWTH RATE")
+        if index != -1:
+            growth_rate.append(float(line.strip().split("=")[1]))
+            continue
+        index = line.find("DELTAW")
+        if index != -1:
+            line_num = k + 4  # table values start 4 lines below heading
+            continue
+        if line_num > 0 and k == line_num:
+            row = line.strip().split()
+            delta_W.append(float(row[2]))
+            surf_idx = int(row[0])
+            line_num = line_num + 1
+            if surf_idx >= surfs - 1:
+                break  # NOTE: not including vacuum surfaces
+    file.close()
+
+    errorif(len(growth_rate) == 0, RuntimeError, "Growth rate not found!")
+    errorif(
+        len(growth_rate) > 1,
+        NotImplementedError,
+        "Not capable of handling multiple growth rates.",
+    )
+    errorif(len(delta_W) != surfs - 1, RuntimeError, "Incorrect number of dW values!")
+
+    return growth_rate[0] if scalar else delta_W
+
+
+class TERPSICHORE(ExternalObjective):
+    """Computes ideal MHD linear stability from calls to the code TERPSICHORE.
+
+    Returns the linear growth rate of the fastest growing instability, or the ΔW values
+    at each flux surface. A positive growth rate or negative ΔW denotes instability.
+
```

Comment:
might be good to cite it as well

https://link.springer.com/chapter/10.1007/978-1-4613-0659-7_8